topcoffea/modules/sparseHist.py

```diff
@@ -1,409 +1,297 @@
-        kwargs: Same as for hist.Hist
+class SparseState:
+    def __init__(self, category_axes, dense_axes, hist_cls):
         """
-
-        self._init_args = dict(kwargs)
-
-        categorical_axes, dense_axes = self._check_args(axes)
-
-        self._tuple_t = namedtuple(
-            f"SparseHistTuple{id(self)}", [a.name for a in categorical_axes]
-        )
-        self._dense_hists: dict[self._tuple_t, hist.Hist] = {}
-
-        # we use self to keep track of the bins in the categorical axes.
-        super().__init__(*categorical_axes, storage="Double")
-
-        self._categorical_axes = super().axes
-        self._dense_axes = hist.axis.NamedAxesTuple(dense_axes)
-
-        self.axes = hist.axis.NamedAxesTuple(chain(super().axes, dense_axes))
-
-    def _check_args(self, axes):
-        on_cats = True
-        categorical_axes = []
-        dense_axes = []
-
-        for axis in axes:
-            if isinstance(axis, (hist.axis.StrCategory, hist.axis.IntCategory)):
-                if not on_cats:
-                    ValueError("All categorical axes should be specified first.")
-                categorical_axes.append(axis)
-            else:
-                on_cats = False
-                dense_axes.append(axis)
-
-        if len(dense_axes) < 1:
-            raise ValueError("At least one dense axis should be specified.")
-
-        return categorical_axes, dense_axes
-
-    def empty_from_axes(self, categorical_axes=None, dense_axes=None, **kwargs):
-        """Create an empty histogram like the current one, but with the axes provided.
-        If axes are None, use those of current histogram.
+        category_axes: List of categorical axes. They should be of type hist.axes.* (i.e., not dask-aware).
+        dense_axes: List of dense axes. Either all or none may be dask-aware.
+        hist_cls: Type of histogram (e.g. hist.Hist or hist.dask.Hist)
         """
-        if categorical_axes is None:
-            categorical_axes = self.categorical_axes
+        self._dense_axes = list(dense_axes)
+        self.dense_hists = defaultdict(lambda: self.make_dense())
+        self.hist_cls = hist_cls
+        self._bookkeep = hist.Hist(*category_axes)
 
-        if dense_axes is None:
-            dense_axes = self.dense_axes
+    @property
+    def category_axes(self):
+        return self._bookkeep.axes
 
-        return type(self)(*categorical_axes, *dense_axes, **kwargs, **self._init_args)
+    @property
+    def dense_axes(self):
+        for a in self._dense_axes:
+            yield a
 
-    def make_dense(self, *axes, **kwargs):
-        return hist.Hist(*axes, **self._init_args, **kwargs)
+    @property
+    def axes(self):
+        return []
 
-    def __copy__(self):
-        """Empty histograms with the same bins."""
-        return self.empty_from_axes(categorical_axes=self.categorical_axes)
+    @property
+    def n_categories(self):
+        return len(self._bookkeep.axes)
```

Comment:
Will `len()` automatically get "daskified"?

Reply:
It does not. But in this case it is not needed. Category axes are "normal" axes which values are available without the need of `dask.compute`. Only the fills for dense axes need the `dask.compute` call.

Reply:
Ok, thanks for clarifying.